lib/AFP/AbstractFilePath/Internal/Common.hs

```diff
@@ -0,0 +1,692 @@
+
+
+takeExtensions :: ShortByteString -> ShortByteString
+takeExtensions = snd . splitExtensions
+
+
+stripExtension :: ShortByteString -> ShortByteString -> Maybe ShortByteString
+stripExtension bs path
+  | BS.null bs = Just path
+  | otherwise  = BS.stripSuffix dotExt path
+  where
+    dotExt = if isExtSeparator $ BS.head bs
+                then bs
+                else extSeparator `BS.cons` bs
+
+
+------------------------
+-- Filename/directory functions
+
+
+splitFileName :: ShortByteString -> (ShortByteString, ShortByteString)
+splitFileName x = if BS.null path
+    then (dotSlash, file)
+    else (path, file)
+  where
+    (path, file) = splitFileNameRaw x
+    dotSlash = _period `BS.cons` (BS.singleton pathSeparator)
+
+
+takeFileName :: ShortByteString -> ShortByteString
+takeFileName = snd . splitFileName -- TODO: null filename not allowed
+
+
+replaceFileName :: ShortByteString -> ShortByteString -> ShortByteString
+replaceFileName x y = fst (splitFileNameRaw x) </> y
+
+
+dropFileName :: ShortByteString -> ShortByteString
+dropFileName = fst . splitFileName
+
+
+takeBaseName :: ShortByteString -> ShortByteString
+takeBaseName = dropExtension . takeFileName
+
+
+replaceBaseName :: ShortByteString -> ShortByteString -> ShortByteString
+replaceBaseName path name = combineRaw dir (name <.> ext)
+  where
+    (dir,file) = splitFileNameRaw path
+    ext = takeExtension file
+
+
+takeDirectory :: ShortByteString -> ShortByteString
+takeDirectory x = case () of
+    () | isPathSeparator' x -> x
+       | BS.null res && not (BS.null file) -> file
+       | otherwise -> res
+  where
+    res = fst $ BS.spanEnd isPathSeparator file
+    file = dropFileName x
+
+
+replaceDirectory :: ShortByteString -> ShortByteString -> ShortByteString
+replaceDirectory file dir = combineRaw dir (takeFileName file)
+
+
+combine :: ShortByteString -> ShortByteString -> ShortByteString
+combine a b | not (BS.null b) && isPathSeparator (BS.head b) || hasDrive b = b
+            | otherwise = combineRaw a b
+
+
+(</>) :: ShortByteString -> ShortByteString -> ShortByteString
+(</>) = combine
+
+splitPath :: ShortByteString -> [ShortByteString]
+splitPath x = [drive | not (BS.null drive)] ++ f path
+    where
+        (drive, path) = splitDrive x
+
+        f y
+          | BS.null y = []
+          | otherwise = (a `BS.append` c) : f d
+            where
+                (a,b) = BS.break isPathSeparator y
+                (c,d) = BS.span  isPathSeparator b
+
+
+joinPath :: [ShortByteString] -> ShortByteString
+joinPath = foldr (</>) mempty
+
+
+splitDirectories :: ShortByteString -> [ShortByteString]
+splitDirectories x
+    | BS.null x = []
+    | isPathSeparator (BS.head x) = let (root,rest) = BS.splitAt 1 x
+                                    in root : splitter rest
+    | otherwise = splitter x
+  where
+    splitter = filter (not . BS.null) . BS.splitWith isPathSeparator
+
+
+takeAllParents :: ShortByteString -> [ShortByteString]
```

Comment:
This function is new and doesn't exist anywhere. I found it useful to emulate e.g. `mkdir -p`.

```
-- > Windows: takeAllParents "C:\\foo\\bar" == ["C:\foo","C:"]
-- > Windows: takeAllParents "/abs/def/dod" == ["\abs\def","\abs","\"]
-- > Windows: takeAllParents "/foo" == ["\"]
-- > Posix: takeAllParents "/abs/def/dod" == ["/abs/def","/abs","/"]
-- > Posix: takeAllParents "/foo" == ["/"]
-- > takeAllParents "/" == []
```